### PR TITLE
experimental: add post-processing logic

### PR DESCRIPTION
**Pull request by DavidKorczynski**
This is needed to easily digest the results from a larger scale run

### CODE REVIEW

**Comment by DavidKorczynski** [comment]
Sample run:
```
$ cat ./targets.txt 
https://github.com/gregjesl/simpleson
https://github.com/memononen/nanosvg
https://github.com/codeplea/tinyexpr
$ python3 ./runner.py -o ../../../oss-fuzz-3/ -i ./targets.txt -t 6 -m openai
....
...
$ python3 ./post-process.py ../../../oss-fuzz-3                                                        
[post-process.py:163]: Analysing: ../../../oss-fuzz-3/build/out/temp-project-0
[post-process.py:174]: Results: https://github.com/gregjesl/simpleson
[post-process.py:175]: - Language: c++
[post-process.py:176]: - Functions from Fuzz Introspector: 108
[post-process.py:178]: - Harnesses generated: 7
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-2-fuzzgen-1/fuzz-run.err :: 18 :: 221 :: [asan errors: 1] :: coverage: 28.930818 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-1-fuzzgen-8/fuzz-run.err :: 14 :: 216 :: [asan errors: 1] :: coverage: 31.238779 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-1-fuzzgen-1/fuzz-run.err :: 18 :: 196 :: [asan errors: 1] :: coverage: 26.145553 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-0-fuzzgen-1/fuzz-run.err :: 18 :: 160 :: [asan errors: 1] :: coverage: 23.899371 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-0-fuzzgen-8/fuzz-run.err :: 14 :: 159 :: [asan errors: 1] :: coverage: 24.955117 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-2-fuzzgen-8/fuzz-run.err :: 14 :: 105 :: [asan errors: 1] :: coverage: 29.174147 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-0/autogen-results/test-fuzz-build-2-fuzzgen-0/fuzz-run.err :: 2 :: 10 :: [asan errors: 0] :: coverage: 2.066487 
[post-process.py:163]: Analysing: ../../../oss-fuzz-3/build/out/temp-project-1
[post-process.py:174]: Results: https://github.com/memononen/nanosvg
[post-process.py:175]: - Language: c
[post-process.py:176]: - Functions from Fuzz Introspector: 173
[post-process.py:178]: - Harnesses generated: 16
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-1-fuzzgen-1/fuzz-run.err :: 25 :: 862 :: [asan errors: 0] :: coverage: 55.141727 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-1-fuzzgen-7/fuzz-run.err :: 2 :: 818 :: [asan errors: 0] :: coverage: 39.282186 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-0-fuzzgen-7/fuzz-run.err :: 2 :: 816 :: [asan errors: 0] :: coverage: 47.836326 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-2-fuzzgen-1/fuzz-run.err :: 25 :: 746 :: [asan errors: 0] :: coverage: 38.991430 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-2-fuzzgen-13/fuzz-run.err :: 32 :: 735 :: [asan errors: 0] :: coverage: 55.511551 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-3-fuzzgen-7/fuzz-run.err :: 2 :: 726 :: [asan errors: 0] :: coverage: 50.576227 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-0-fuzzgen-1/fuzz-run.err :: 25 :: 717 :: [asan errors: 0] :: coverage: 36.564526 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-0-fuzzgen-13/fuzz-run.err :: 32 :: 703 :: [asan errors: 0] :: coverage: 38.289086 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-1-fuzzgen-13/fuzz-run.err :: 32 :: 657 :: [asan errors: 0] :: coverage: 44.092409 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-2-fuzzgen-7/fuzz-run.err :: 2 :: 648 :: [asan errors: 0] :: coverage: 41.389529 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-3-fuzzgen-13/fuzz-run.err :: 32 :: 628 :: [asan errors: 0] :: coverage: 39.834983 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-3-fuzzgen-1/fuzz-run.err :: 25 :: 624 :: [asan errors: 0] :: coverage: 41.957811 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-0-fuzzgen-0/fuzz-run.err :: 2 :: 358 :: [asan errors: 0] :: coverage: 26.456739 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-3-fuzzgen-0/fuzz-run.err :: 2 :: 328 :: [asan errors: 0] :: coverage: 37.590520 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-2-fuzzgen-0/fuzz-run.err :: 2 :: 302 :: [asan errors: 0] :: coverage: 27.649770 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-1/autogen-results/test-fuzz-build-1-fuzzgen-0/fuzz-run.err :: 2 :: 284 :: [asan errors: 0] :: coverage: 31.566820 
[post-process.py:163]: Analysing: ../../../oss-fuzz-3/build/out/temp-project-2
[post-process.py:174]: Results: https://github.com/codeplea/tinyexpr
[post-process.py:175]: - Language: c
[post-process.py:176]: - Functions from Fuzz Introspector: 65
[post-process.py:178]: - Harnesses generated: 14
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-15/fuzz-run.err :: 30 :: 180 :: [asan errors: 0] :: coverage: 34.369115 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-15/fuzz-run.err :: 30 :: 170 :: [asan errors: 0] :: coverage: 33.081285 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-16/fuzz-run.err :: 28 :: 169 :: [asan errors: 0] :: coverage: 33.084809 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-21/fuzz-run.err :: 30 :: 168 :: [asan errors: 0] :: coverage: 34.463277 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-16/fuzz-run.err :: 28 :: 162 :: [asan errors: 0] :: coverage: 33.208606 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-4/fuzz-run.err :: 28 :: 161 :: [asan errors: 0] :: coverage: 32.645403 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-4/fuzz-run.err :: 28 :: 161 :: [asan errors: 0] :: coverage: 32.710280 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-21/fuzz-run.err :: 30 :: 159 :: [asan errors: 0] :: coverage: 34.146341 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-3/fuzz-run.err :: 30 :: 152 :: [asan errors: 0] :: coverage: 34.277620 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-9/fuzz-run.err :: 30 :: 152 :: [asan errors: 0] :: coverage: 32.956685 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-3/fuzz-run.err :: 30 :: 151 :: [asan errors: 0] :: coverage: 33.019755 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-9/fuzz-run.err :: 30 :: 151 :: [asan errors: 0] :: coverage: 34.310019 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-1-fuzzgen-22/fuzz-run.err :: 28 :: 50 :: [asan errors: 0] :: coverage: 32.833021 
[post-process.py:180]: ../../../oss-fuzz-3/build/out/temp-project-2/autogen-results/test-fuzz-build-2-fuzzgen-22/fuzz-run.err :: 28 :: 43 :: [asan errors: 0] :: coverage: 32.710280 
```

File: experimental/c-cpp/post-process.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  fuzz_run = os.path.join(generated_sample_setup_dir, 'fuzz-run.err')
+  if not os.path.isfile(fuzz_run):
+    return None
+
+  run_results = interpret_llvm_run(fuzz_run)
+  run_results['target'] = fuzz_run
+
+  # Check the covreage run
+  coverage_proj = os.path.basename(
+      project_folder_out) + "-cov-" + generated_sample_setup
```
Comment (DonggeLiu) [inline experimental/c-cpp/post-process.py:93]: nit: Single quote for consistency.

File: experimental/c-cpp/post-process.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+def extract_results_from_generated_samples(
+    oss_fuzz_dir: str, project_folder_out: str) -> List[Dict[str, Any]]:
+  """Iterates an auto-gen session and returns a list of successful results."""
+  total_outcomes = []
+  for generated_sample_setup in os.listdir(
+      os.path.join(project_folder_out, constants.SHARED_MEMORY_RESULTS_DIR)):
+    generated_sample_setup_dir = os.path.join(
+        project_folder_out, constants.SHARED_MEMORY_RESULTS_DIR,
+        generated_sample_setup)
+    #print("Analysing %s"%(dd))
```
Comment (DonggeLiu) [inline experimental/c-cpp/post-process.py:119]: Remove debug print, this was logged at another place.

**Review by DonggeLiu — APPROVED** [review]

File: experimental/c-cpp/post-process.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+  for project_out in os.listdir(oss_fuzz_out_dir(oss_fuzz_dir)):
+    if project_out.startswith(constants.PROJECT_BASE):
+      try:
+        project_indices.append(
+            int(project_out.replace(constants.PROJECT_BASE, '')))
+      except ValueError:
+        pass
+
+  sorted_project_folders = [
+      constants.PROJECT_BASE + '%d' % (idx)
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/post-process.py]: done, thanks

File: experimental/c-cpp/post-process.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  for line_idx, word in enumerate(split_line):
+    if word == 'cov:':
+      try:
+        return int(split_line[line_idx + 1])
+      except (ValueError, IndexError):
+        pass
+  return None
+
+
+def interpret_llvm_run(log_path: str) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/post-process.py]: done

File: experimental/c-cpp/post-process.py (hunk 2 above)
Comment (DavidKorczynski) [reply experimental/c-cpp/post-process.py]: done

File: experimental/c-cpp/post-process.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        cov_history.append(edge_coverage)
+      if 'ERROR: AddressSanitizer' in line:
+        addr_sanitizer_lines.append(line)
+
+  return {'cov_history': cov_history, 'addr_san': addr_sanitizer_lines}
+
+
+def oss_fuzz_out_dir(oss_fuzz_dir: str) -> str:
+  """Returns the out folder of a given OSS-Fuzz repository."""
+  return os.path.join(oss_fuzz_dir, 'build', 'out')
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/post-process.py:58]: Good point. I'll leave this for now as this PR is in a bit grander-scheme an effort to prepare for merging the codebases https://github.com/google/oss-fuzz-gen/issues/450 -- A larger PR will come in the near future that does this on a more rigorous level so will do this at that point

File: experimental/c-cpp/post-process.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+import sys
+from typing import Any, Dict, List, Optional
+
+import constants
+
+logger = logging.getLogger(name=__name__)
+LOG_FMT = '[%(filename)s:%(lineno)d]: %(message)s'
+
+
+def _get_edge_cov_from_line(line: str) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/post-process.py:28]: Will do this in a similar fashion as mentioned in https://github.com/google/oss-fuzz-gen/pull/440#discussion_r1668300582

**Pull request merged**